File: c/difftest.c
/*
 * difftest.c - differential harness between the C port and the Rust reference.
 *
 * Reads lines of the form
 *
 *   g,input_hex,ctx_hex,out_hex
 *
 * produced by `cargo run --release --example difftest` (see ../examples/
 * difftest.rs). For every case it derives FPR once, ratchets the two chains
 * from (S_0, K_r) to generation g, re-encodes the input with fre_encode, and
 * compares byte-for-byte against out_hex. Any mismatch or length anomaly
 * exits non-zero with diagnostics.
 *
 * Uses the same pinned test constants as c/tests.c and the vectors in the
 * README, so this doubles as a chain-agreement check: if the Rust and C
 * rotations ever diverge at some generation, every encode at or past it
 * mismatches here.
 */

#define _POSIX_C_SOURCE 200809L /* getline / ssize_t under -std=c99 */

#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "fre.h"

static const uint8_t S0[32] = {0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07,
                               0x08, 0x09, 0x0a, 0x0b, 0x0c, 0x0d, 0x0e, 0x0f,
                               0x10, 0x11, 0x12, 0x13, 0x14, 0x15, 0x16, 0x17,
                               0x18, 0x19, 0x1a, 0x1b, 0x1c, 0x1d, 0x1e, 0x1f};

static const uint8_t KR[32] = {0x2a, 0x2a, 0x2a, 0x2a, 0x2a, 0x2a, 0x2a, 0x2a,
                               0x2a, 0x2a, 0x2a, 0x2a, 0x2a, 0x2a, 0x2a, 0x2a,
                               0x2a, 0x2a, 0x2a, 0x2a, 0x2a, 0x2a, 0x2a, 0x2a,
                               0x2a, 0x2a, 0x2a, 0x2a, 0x2a, 0x2a, 0x2a, 0x2a};

static int hexval(int c) {
  if (c >= '0' && c <= '9')
    return c - '0';
  if (c >= 'a' && c <= 'f')
    return c - 'a' + 10;
  if (c >= 'A' && c <= 'F')
    return c - 'A' + 10;
  return -1;
}

static size_t decode_hex(const char *s, uint8_t **out) {
  size_t len = strlen(s), bytes, i;
  uint8_t *buf;
  if (len % 2) {
    fprintf(stderr, "difftest: odd hex length in field '%s'\n", s);
    exit(3);
  }
  bytes = len / 2;
  buf = malloc(bytes ? bytes : 1);
  if (!buf) {
    fprintf(stderr, "difftest: out of memory\n");
    exit(3);
  }
  for (i = 0; i < bytes; i++) {
    int hi = hexval((unsigned char)s[2 * i]);
    int lo = hexval((unsigned char)s[2 * i + 1]);
    if (hi < 0 || lo < 0) {
      fprintf(stderr, "difftest: bad hex in field '%.64s'\n", s);
      exit(3);
    }
    buf[i] = (uint8_t)((hi << 4) | lo);
  }
  *out = buf;
  return bytes;
}

static void to_hex(const uint8_t *b, size_t n, char *out) {
  static const char d[] = "0123456789abcdef";
  size_t i;
  for (i = 0; i < n; i++) {
    out[2 * i] = d[b[i] >> 4];
    out[2 * i + 1] = d[b[i] & 0x0f];
  }
  out[2 * n] = 0;
}

int main(void) {
  char *line = NULL;
  size_t cap = 0;
  ssize_t n;
  uint8_t fpr[32];
  uint64_t cases = 0;

  fre_fpr(S0, KR, fpr);

  while ((n = getline(&line, &cap, stdin)) >= 0) {
    uint64_t g;
    char *g_str, *in_hex, *ctx_hex, *out_hex, *c1, *c2, *c3;
    uint8_t *in = NULL, *c = NULL, *outb = NULL, *got;
    size_t in_len, ctx_len, out_len, expect, got_n;
    uint8_t sg[32], kg[32];
    char *got_hex;
    uint64_t gi;

    while (n > 0 && (line[n - 1] == '\n' || line[n - 1] == '\r')) {
      line[--n] = 0;
    }
    if (n == 0) {
      continue;
    }

    g_str = line;
    c1 = strchr(line, ',');
    if (!c1) {
      fprintf(stderr, "difftest: no generation field\n");
      return 2;
    }
    *c1 = 0;
    in_hex = c1 + 1;
    c2 = strchr(in_hex, ',');
    if (!c2) {
      fprintf(stderr, "difftest: no input field\n");
      return 2;
    }
    *c2 = 0;
    ctx_hex = c2 + 1;
    c3 = strchr(ctx_hex, ',');
    if (!c3) {
      fprintf(stderr, "difftest: no ctx field\n");
      return 2;
    }
    *c3 = 0;
    out_hex = c3 + 1;

    g = strtoull(g_str, NULL, 10);
    in_len = decode_hex(in_hex, &in);
    ctx_len = decode_hex(ctx_hex, &c);
    out_len = decode_hex(out_hex, &outb);
    if (ctx_len > FRE_CTX_MAX) {
      fprintf(stderr, "difftest: ctx_len %zu exceeds FRE_CTX_MAX\n", ctx_len);
      return 2;
    }

    memcpy(sg, S0, 32);
    memcpy(kg, KR, 32);
    for (gi = 0; gi < g; gi++) {
      fre_rotate(sg, kg, fpr, gi, sg, kg);
    }

    expect = fre_out_len(in_len);
    if (expect != out_len) {
      fprintf(stderr,
              "difftest: case %llu g=%llu in_len=%zu: C output length %zu "
              "!= Rust output length %zu\n",
              (unsigned long long)cases, (unsigned long long)g, in_len, expect,
              out_len);
      return 1;
    }

    got = malloc(expect ? expect : 1);
    if (!got) {
      fprintf(stderr, "difftest: out of memory\n");
      return 3;
    }
    got_n = fre_encode(in, in_len, sg, fpr, (uint64_t)g,
                       ctx_len ? c : NULL, ctx_len, got);
    if (got_n != expect) {
      fprintf(stderr, "difftest: case %llu: fre_encode returned %zu, want %zu\n",
              (unsigned long long)cases, got_n, expect);
      return 1;
    }

    got_hex = malloc(expect * 2 + 1);
    if (!got_hex) {
      fprintf(stderr, "difftest: out of memory\n");
      return 3;
    }
    to_hex(got, expect, got_hex);
    if (strcmp(got_hex, out_hex) != 0) {
      fprintf(stderr,
              "MISMATCH case %llu: g=%llu in_len=%zu ctx_len=%zu\n"
              "  rust out_hex = %.80s...\n"
              "  c    out_hex = %.80s...\n",
              (unsigned long long)cases, (unsigned long long)g, in_len,
              ctx_len, out_hex, got_hex);
      return 1;
    }

    cases++;
    free(in);
    free(c);
    free(outb);
    free(got);
    free(got_hex);
  }

  free(line);
  printf("difftest: %llu cases OK\n", (unsigned long long)cases);
  return 0;
}

File: examples/difftest.rs
//! Differential test-vector generator (Rust side).
//!
//! Emits deterministic FRE cases as `g,input_hex,ctx_hex,out_hex` lines fed to
//! `c/difftest` (and `c/difftest_asan`), which recompute every case with the C
//! port. Any divergence between the two implementations fails the C side.
//!
//! The case mix is aimed at the boundaries that matter:
//!   - input lengths 0..160 dense (covers the 63/64/65 and 127/128/129
//!     padding cliffs), then random lengths up to 4096 bytes;
//!   - ctx lengths concentrated on 0,7,9,16,17,40,63,64 plus random, so the
//!     HMAC key crosses 64 bytes (ctx 17 -> key 65) and the FRE_CTX_MAX bound
//!     (ctx 65 is never emitted; violations are exercised in the unit tests);
//!   - g concentrated on 0,1,41,42,43 plus random 0..=64, so both chains are
//!     rotated far from bootstrap in a reproducible way.
//!
//! PRNG is xorshift64* with a fixed default seed (no external deps); pass a
//! different `[N] [seed]` to re-roll while staying reproducible.

use std::env;

use rust_fre::{CTX_MAX, fpr, fre, rotate};

const S0: [u8; 32] = *b"\x00\x01\x02\x03\x04\x05\x06\x07\x08\x09\x0a\x0b\x0c\x0d\x0e\x0f\x10\x11\x12\x13\x14\x15\x16\x17\x18\x19\x1a\x1b\x1c\x1d\x1e\x1f";
const KR: [u8; 32] = [0x2a; 32];

struct XorShift(u64);

impl XorShift {
    fn next_u64(&mut self) -> u64 {
        let mut x = self.0;
        x ^= x >> 12;
        x ^= x << 25;
        x ^= x >> 27;
        self.0 = x;
        x.wrapping_mul(0x2545_F491_4F6C_DD1D)
    }

    fn below(&mut self, n: u64) -> u64 {
        self.next_u64() % n
    }

    fn byte(&mut self) -> u8 {
        (self.next_u64() >> 56) as u8
    }
}

fn parse_u64(s: &str) -> u64 {
    if let Some(hex) = s.strip_prefix("0x") {
        u64::from_str_radix(hex, 16).expect("hex seed")
    } else {
        s.parse().expect("decimal seed")
    }
}

fn hex(b: &[u8]) -> String {
    let mut out = String::with_capacity(b.len() * 2);
    for x in b {
        out.push_str(&format!("{x:02x}"));
    }
    out
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let n: usize = args
        .get(1)
        .map(|a| a.parse().expect("usage: difftest [N] [seed]"))
        .unwrap_or(20_000);
    let mut seed = args
        .get(2)
        .map(|a| parse_u64(a))
        .unwrap_or(0xa5a5_a5a5_a5a5_a5a5);
    if seed == 0 {
        seed = 1; // xorshift64* is stuck on state 0
    }
    let mut r = XorShift(seed);

    let f = fpr(&S0, &KR);

    // Precompute the two chains once: S_g and K_g for g in 0..=MAXG.
    const MAXG: usize = 64;
    let mut seeds = [[0u8; 32]; MAXG + 1];
    let mut keys = [[0u8; 32]; MAXG + 1];
    seeds[0] = S0;
    keys[0] = KR;
    for g in 0..MAXG {
        (seeds[g + 1], keys[g + 1]) = rotate(&seeds[g], &keys[g], &f, g as u64);
    }

    const SPECIAL_CTX: [usize; 8] = [0, 7, 9, 16, 17, 40, 63, 64];
    const SPECIAL_G: [u64; 5] = [0, 1, 41, 42, 43];

    for i in 0..n {
        let g = if r.below(6) == 0 {
            SPECIAL_G[r.below(SPECIAL_G.len() as u64) as usize]
        } else {
            r.below(MAXG as u64 + 1)
        };
        let ctx_len = if r.below(5) == 0 {
            SPECIAL_CTX[r.below(SPECIAL_CTX.len() as u64) as usize]
        } else {
            r.below(CTX_MAX as u64 + 1) as usize
        };
        let in_len = if i < 160 {
            i
        } else if r.below(10) < 7 {
            r.below(301) as usize
        } else {
            r.below(4097) as usize
        };
        let mut input = vec![0u8; in_len];
        for b in &mut input {
            *b = r.byte();
        }
        let mut ctx = vec![0u8; ctx_len];
        for b in &mut ctx {
            *b = r.byte();
        }
        let out = fre(&input, &seeds[g as usize], &f, g, &ctx);
        println!("{g},{},{},{}", hex(&input), hex(&ctx), hex(&out));
    }
}

